// @wdio/cucumber-framework/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.When = exports.Then = exports.setWorldConstructor = exports.setDefinitionFunctionWrapper = exports.setDefaultTimeout = exports.Given = exports.defineStep = exports.defineParameterType = exports.BeforeStep = exports.BeforeAll = exports.Before = exports.AfterStep = exports.AfterAll = exports.After = exports.adapterFactory = exports.CucumberAdapter = void 0;
const path_1 = __importDefault(require("path"));
const events_1 = require("events");
const mockery_1 = __importDefault(require("mockery"));
const is_glob_1 = __importDefault(require("is-glob"));
const glob_1 = __importDefault(require("glob"));
const Cucumber = __importStar(require("@cucumber/cucumber"));
const cucumber_1 = require("@cucumber/cucumber");
Object.defineProperty(exports, "After", { enumerable: true, get: function () { return cucumber_1.After; } });
Object.defineProperty(exports, "AfterAll", { enumerable: true, get: function () { return cucumber_1.AfterAll; } });
Object.defineProperty(exports, "AfterStep", { enumerable: true, get: function () { return cucumber_1.AfterStep; } });
Object.defineProperty(exports, "Before", { enumerable: true, get: function () { return cucumber_1.Before; } });
Object.defineProperty(exports, "BeforeAll", { enumerable: true, get: function () { return cucumber_1.BeforeAll; } });
Object.defineProperty(exports, "BeforeStep", { enumerable: true, get: function () { return cucumber_1.BeforeStep; } });
Object.defineProperty(exports, "defineParameterType", { enumerable: true, get: function () { return cucumber_1.defineParameterType; } });
Object.defineProperty(exports, "defineStep", { enumerable: true, get: function () { return cucumber_1.defineStep; } });
Object.defineProperty(exports, "Given", { enumerable: true, get: function () { return cucumber_1.Given; } });
Object.defineProperty(exports, "setDefaultTimeout", { enumerable: true, get: function () { return cucumber_1.setDefaultTimeout; } });
Object.defineProperty(exports, "setDefinitionFunctionWrapper", { enumerable: true, get: function () { return cucumber_1.setDefinitionFunctionWrapper; } });
Object.defineProperty(exports, "setWorldConstructor", { enumerable: true, get: function () { return cucumber_1.setWorldConstructor; } });
Object.defineProperty(exports, "Then", { enumerable: true, get: function () { return cucumber_1.Then; } });
Object.defineProperty(exports, "When", { enumerable: true, get: function () { return cucumber_1.When; } });
const gherkin_streams_1 = require("@cucumber/gherkin-streams");
const event_data_collector_1 = __importDefault(require("@cucumber/cucumber/lib/formatter/helpers/event_data_collector"));
const messages_1 = require("@cucumber/messages");
const utils_1 = require("@wdio/utils");
const reporter_1 = __importDefault(require("./reporter"));
const constants_1 = require("./constants");
const utils_2 = require("./utils");
const { incrementing } = messages_1.IdGenerator;
function getResultObject(world) {
    var _a, _b, _c, _d;
    return {
        passed: ((_a = world.result) === null || _a === void 0 ? void 0 : _a.status) === Cucumber.Status.PASSED,
        error: (_b = world.result) === null || _b === void 0 ? void 0 : _b.message,
        duration: ((_d = (_c = world.result) === null || _c === void 0 ? void 0 : _c.duration) === null || _d === void 0 ? void 0 : _d.nanos) / 10e6 // convert into ms
    };
}
class CucumberAdapter {
    constructor(_cid, _config, _specs, _capabilities, _reporter) {
        this._cid = _cid;
        this._config = _config;
        this._specs = _specs;
        this._capabilities = _capabilities;
        this._reporter = _reporter;
        this._cwd = process.cwd();
        this._newId = incrementing();
        this._cucumberOpts = Object.assign({}, constants_1.DEFAULT_OPTS, this._config.cucumberOpts);
        this._hasTests = true;
        this._cucumberFeaturesWithLineNumbers = this._config.cucumberFeaturesWithLineNumbers || [];
        this._eventBroadcaster = new events_1.EventEmitter();
        this._eventDataCollector = new event_data_collector_1.default(this._eventBroadcaster);
        const featurePathsToRun = this._cucumberFeaturesWithLineNumbers.length > 0 ? this._cucumberFeaturesWithLineNumbers : this._specs;
        this._pickleFilter = new Cucumber.PickleFilter({
            cwd: this._cwd,
            featurePaths: featurePathsToRun,
            names: this._cucumberOpts.names,
            tagExpression: this._cucumberOpts.tagExpression
        });
        const reporterOptions = {
            capabilities: this._capabilities,
            ignoreUndefinedDefinitions: Boolean(this._cucumberOpts.ignoreUndefinedDefinitions),
            failAmbiguousDefinitions: Boolean(this._cucumberOpts.failAmbiguousDefinitions),
            tagsInTitle: Boolean(this._cucumberOpts.tagsInTitle),
            scenarioLevelReporter: Boolean(this._cucumberOpts.scenarioLevelReporter)
        };
        this._cucumberReporter = new reporter_1.default(this._eventBroadcaster, this._pickleFilter, reporterOptions, this._cid, this._specs, this._reporter);
    }
    async init() {
        try {
            const gherkinMessageStream = gherkin_streams_1.GherkinStreams.fromPaths(this._specs, {
                defaultDialect: this._cucumberOpts.featureDefaultLanguage,
                newId: this._newId
            });
            await Cucumber.parseGherkinMessageStream({
                cwd: this._cwd,
                eventBroadcaster: this._eventBroadcaster,
                gherkinMessageStream,
                eventDataCollector: this._eventDataCollector,
                order: this._cucumberOpts.order,
                pickleFilter: this._pickleFilter
            });
            this._hasTests = this._cucumberReporter.eventListener.getPickleIds(this._capabilities).length > 0;
        }
        catch (runtimeError) {
            await (0, utils_1.executeHooksWithArgs)('after', this._config.after, [runtimeError, this._capabilities, this._specs]);
            throw runtimeError;
        }
        /**
         * import and set options for `expect-webdriverio` assertion lib once
         * the framework was initiated so that it can detect the environment
         */
        const { setOptions } = require('expect-webdriverio');
        setOptions({
            wait: this._config.waitforTimeout,
            interval: this._config.waitforInterval, // interval between attempts
        });
        return this;
    }
    hasTests() {
        return this._hasTests;
    }
    async run() {
        let runtimeError;
        let result;
        try {
            this.registerRequiredModules();
            Cucumber.supportCodeLibraryBuilder.reset(this._cwd, this._newId);
            /**
             * wdio hooks should be added before spec files are loaded
             */
            this.addWdioHooks(this._config);
            this.loadSpecFiles();
            this.wrapSteps(this._config);
            /**
             * we need to somehow identify is function is step or hook
             * so we wrap every user hook function
             */
            (0, utils_2.setUserHookNames)(Cucumber.supportCodeLibraryBuilder);
            Cucumber.setDefaultTimeout(this._cucumberOpts.timeout);
            const supportCodeLibrary = Cucumber.supportCodeLibraryBuilder.finalize();
            /**
             * gets current step data: `{ uri, feature, scenario, step, sourceLocation }`
             * or `null` for some hooks.
             */
            this.getHookParams = this._cucumberReporter
                .eventListener
                .getHookParams
                .bind(this._cucumberReporter.eventListener);
            const runtime = new Cucumber.Runtime({
                newId: this._newId,
                eventBroadcaster: this._eventBroadcaster,
                options: this._cucumberOpts,
                supportCodeLibrary,
                eventDataCollector: this._eventDataCollector,
                pickleIds: this._cucumberReporter.eventListener.getPickleIds(this._capabilities)
            });
            result = await runtime.start() ? 0 : 1;
            /**
             * if we ignore undefined definitions we trust the reporter
             * with the fail count
             */
            if (this._cucumberOpts.ignoreUndefinedDefinitions && result) {
                result = this._cucumberReporter.failedCount;
            }
        }
        catch (err) {
            runtimeError = err;
            result = 1;
        }
        await (0, utils_1.executeHooksWithArgs)('after', this._config.after, [runtimeError || result, this._capabilities, this._specs]);
        /**
         * in case the spec has a runtime error throw after the wdio hook
         */
        if (runtimeError) {
            throw runtimeError;
        }
        return result;
    }
    /**
     * Transpilation https://github.com/cucumber/cucumber-js/blob/master/docs/cli.md#transpilation
     * Usage: `['module']`
     * we extend it a bit with ability to init and pass configuration to modules.
     * Pass an array with path to module and its configuration instead:
     * Usage: `[['module', {}]]`
     * Or pass your own function
     * Usage: `[() => { require('@babel/register')({ ignore: [] }) }]`
     */
    registerRequiredModules() {
        this._cucumberOpts.requireModule.map(requiredModule => {
            if (Array.isArray(requiredModule)) {
                require(requiredModule[0])(requiredModule[1]);
            }
            else if (typeof requiredModule === 'function') {
                requiredModule();
            }
            else {
                require(requiredModule);
            }
        });
    }
    requiredFiles() {
        return this._cucumberOpts.require.reduce((files, requiredFile) => files.concat((0, is_glob_1.default)(requiredFile)
            ? glob_1.default.sync(requiredFile)
            : [requiredFile]), []);
    }
    loadSpecFiles() {
        // we use mockery to allow people to import 'our' cucumber even though their spec files are in their folders
        // because of that we don't have to attach anything to the global object, and the current cucumber spec files
        // should just work with no changes with this framework
        mockery_1.default.enable({
            useCleanCache: false,
            warnOnReplace: false,
            warnOnUnregistered: false
        });
        mockery_1.default.registerMock('@cucumber/cucumber', Cucumber);
        this.requiredFiles().forEach((codePath) => {
            const filepath = path_1.default.isAbsolute(codePath)
                ? codePath
                : path_1.default.join(process.cwd(), codePath);
            // This allows rerunning a stepDefinitions file
            delete require.cache[require.resolve(filepath)];
            require(filepath);
        });
        mockery_1.default.disable();
    }
    /**
     * set `beforeFeature`, `afterFeature`, `beforeScenario`, `afterScenario`, 'beforeStep', 'afterStep'
     * @param {object} config config
     */
    addWdioHooks(config) {
        var _a;
        const eventListener = (_a = this._cucumberReporter) === null || _a === void 0 ? void 0 : _a.eventListener;
        Cucumber.BeforeAll(async function wdioHookBeforeFeature() {
            const params = eventListener === null || eventListener === void 0 ? void 0 : eventListener.getHookParams();
            await (0, utils_1.executeHooksWithArgs)('beforeFeature', config.beforeFeature, [params === null || params === void 0 ? void 0 : params.uri, params === null || params === void 0 ? void 0 : params.feature]);
        });
        Cucumber.AfterAll(async function wdioHookAfterFeature() {
            const params = eventListener === null || eventListener === void 0 ? void 0 : eventListener.getHookParams();
            await (0, utils_1.executeHooksWithArgs)('afterFeature', config.afterFeature, [params === null || params === void 0 ? void 0 : params.uri, params === null || params === void 0 ? void 0 : params.feature]);
        });
        Cucumber.Before(async function wdioHookBeforeScenario(world) {
            await (0, utils_1.executeHooksWithArgs)('beforeScenario', config.beforeScenario, [world, this]);
        });
        Cucumber.After(async function wdioHookAfterScenario(world) {
            await (0, utils_1.executeHooksWithArgs)('afterScenario', config.afterScenario, [world, getResultObject(world), this]);
        });
        Cucumber.BeforeStep(async function wdioHookBeforeStep() {
            const params = eventListener === null || eventListener === void 0 ? void 0 : eventListener.getHookParams();
            await (0, utils_1.executeHooksWithArgs)('beforeStep', config.beforeStep, [params === null || params === void 0 ? void 0 : params.step, params === null || params === void 0 ? void 0 : params.scenario, this]);
        });
        Cucumber.AfterStep(async function wdioHookAfterStep(world) {
            const params = eventListener === null || eventListener === void 0 ? void 0 : eventListener.getHookParams();
            await (0, utils_1.executeHooksWithArgs)('afterStep', config.afterStep, [params === null || params === void 0 ? void 0 : params.step, params === null || params === void 0 ? void 0 : params.scenario, getResultObject(world), this]);
        });
    }
    /**
     * wraps step definition code with sync/async runner with a retry option
     * @param {object} config
     */
    wrapSteps(config) {
        const wrapStep = this.wrapStep;
        const cid = this._cid;
        const getHookParams = () => this.getHookParams && this.getHookParams();
        Cucumber.setDefinitionFunctionWrapper((fn, options = { retry: 0 }) => {
            /**
             * hooks defined in wdio.conf are already wrapped
             */
            if (fn.name.startsWith('wdioHook')) {
                return fn;
            }
            /**
             * this flag is used to:
             * - avoid hook retry
             * - avoid wrap hooks with beforeStep and afterStep
             */
            const isStep = !fn.name.startsWith('userHook');
            return wrapStep(fn, isStep, config, cid, options, getHookParams);
        });
    }
    /**
     * wrap step definition to enable retry ability
     * @param   {Function}  code            step definition
     * @param   {boolean}   isStep
     * @param   {object}    config
     * @param   {string}    cid             cid
     * @param   {StepDefinitionOptions} options
     * @param   {Function}  getHookParams  step definition
     * @return  {Function}                  wrapped step definition for sync WebdriverIO code
     */
    wrapStep(code, isStep, config, cid, options, getHookParams) {
        return function (...args) {
            const hookParams = getHookParams();
            const retryTest = isStep && isFinite(options.retry) ? options.retry : 0;
            /**
             * wrap user step/hook with wdio before/after hooks
             */
            const beforeFn = config.beforeHook;
            const afterFn = config.afterHook;
            return utils_1.testFnWrapper.call(this, isStep ? 'Step' : 'Hook', { specFn: code, specFnArgs: args }, { beforeFn: beforeFn, beforeFnArgs: (context) => [hookParams === null || hookParams === void 0 ? void 0 : hookParams.step, context] }, { afterFn: afterFn, afterFnArgs: (context) => [hookParams === null || hookParams === void 0 ? void 0 : hookParams.step, context] }, cid, retryTest);
        };
    }
}
exports.CucumberAdapter = CucumberAdapter;
const _CucumberAdapter = CucumberAdapter;
const adapterFactory = {};
exports.adapterFactory = adapterFactory;
/**
 * tested by smoke tests
 */
/* istanbul ignore next */
adapterFactory.init = async function (...args) {
    // @ts-ignore just passing through args
    const adapter = new _CucumberAdapter(...args);
    const instance = await adapter.init();
    return instance;
};
exports.default = adapterFactory;
